mvrx/src/main/kotlin/com/airbnb/mvrx/FlowExtensions.kt

```diff
@@ -42,11 +42,14 @@ internal fun <T : Any?> Flow<T>.collectLatest(
         // This is necessary when Dispatchers.Main.immediate is used in scope.
         // Coroutine is launched with start = CoroutineStart.UNDISPATCHED to perform dispatch only once.
         yield()
-        flow.collectLatest {
-            if (MavericksTestOverrides.FORCE_DISABLE_LIFECYCLE_AWARE_OBSERVER) {
-                action(it)
-            } else {
-                lifecycleOwner.whenStarted { action(it) }
+        if (MavericksTestOverrides.FORCE_DISABLE_LIFECYCLE_AWARE_OBSERVER) {
+            flow.collectLatest { action(it) }
+        } else {
+            val lifeCycleAwareDispatcher = lifecycleOwner.lifecycle.lifeCycleAwareDispatcher()
+            flow.collectLatest { item ->
```

Comment:
I feel there is one interesting consequence of this change... 🤔 before, the onEach block was cooperative to cancellation. That means a consumer should decide whether they want or don't want to allow cancellation for the block. 
e.g. 
```
val items = mutableList<Int>()
viewModel.onEach(State::someItem) { item ->
   items.add(item)
}
```
In this example, onEach's body doesn't support cooperative cancellation - the whole body is blocking. If VM emits 1,2,3,4,5. `items` will contain all of them. 

Now, the problem is withContext checks for cancellation, and won't execute the block if it was cancelled (in case if a new item will appear). So effectively it's an equivalent of
```
val items = mutableList<Int>()
viewModel.onEach(State::someItem) { item ->
   currentCoroutineContext.ensureActive()
   items.add(item)
}
```
That means if VM emits 1,2,3,4,5 the behavior is undefined - your `items` may contain a subsequence of these items, e.g. `1,5` or `5` or `1,2,3,4,5`. 

I think it's a huge behavior change that may break a lot of things. IIRC @gpeal was saying they had "conflation" in one of recent versions, but then intentionally avoided it. 

Actually, the same problem is here for pure `lifecycleOwner.whenStarted` call.

I'm trying to think how to change that behavior - outer withContext will definitely help, but it might affect the behavior when lifecycle is paused. Anyway we need a test for this (maybe we have that one? and it somehow passed due to "no dispatch" behavior or smth like that)

Reply:
well you are right about the changing behaviour, but imo this behaviour is more correct as why even bother UI layer with the intermediate state if it's not relevant anymore in case of slow collector?

Wrapping the whole flow collection with paused dispatcher will break `DeliveryMode` I think, values emitted during paused won't be recorded

Reply:
hm, shouldn't UI reset the state when it handles the state rather than VM take a guess if event was handled or not.

But anyway I agree with you about changing compatibility. I will try to think how to address it.

Reply:
looks like wrapping the whole collector into `whenStarted` just works.